README
NAME
    CatalystX::InjectComponent - Inject components into your Catalyst
    application

VERSION
    Version 0.024

SYNOPSIS
        package My::App;

        use Catalyst::Runtime '5.80';

        use Moose;
        BEGIN { extends qw/Catalyst/ }

        ...

        after 'setup_components' => sub {
            my $class = shift;
            CatalystX::InjectComponent->inject( into => $class, component => 'MyModel' );
            if ( $class->config->{ ... ) {
                CatalystX::InjectComponent->inject( into => $class, component => 'MyRootV2', as => 'Controller::Root' );
            }
            else {
                CatalystX::InjectComponent->inject( into => $class, component => 'MyRootV1', as => 'Root' ); # Controller:: will be automatically prefixed
            }
        };

DESCRIPTION
    CatalystX::InjectComponent will inject Controller, Model, and View
    components into your Catalyst application at setup (run)time. It does
    this by creating a new package on-the-fly, having that package extend
    the given component, and then having Catalyst setup the new component
    (via "->setup_component")

So, how do I use this thing?
    You should inject your components when appropiate, typically after
    "setup_compenents" runs

    If you're using the Moose version of Catalyst, then you can use the
    following technique:

        use Moose;
        BEGIN { extends qw/Catalyst/ }

        after 'setup_components' => sub {
            my $class = shift;

            CatalystX::InjectComponent->inject( into => $class, ... )
        };

METHODS
  CatalystX::InjectComponent->inject( ... )
        into        The Catalyst package to inject into (e.g. My::App)
        component   The component package to inject
        as          An optional moniker to use as the package name for the derived component

    For example:

        ->inject( into => My::App, component => Other::App::Controller::Apple )
            
        The above will create 'My::App::Controller::Other::App::Controller::Apple'

        ->inject( into => My::App, component => Other::App::Controller::Apple, as => Apple )

            The above will create 'My::App::Controller::Apple'

ACKNOWLEDGEMENTS
    Inspired by Catalyst::Plugin::AutoCRUD

AUTHOR
    Robert Krimen, "<rkrimen at cpan.org>"

BUGS
    Please report any bugs or feature requests to
    "bug-catalystx-injectcomponent at rt.cpan.org", or through the web
    interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CatalystX-InjectComponen
    t>. I will be notified, and then you'll automatically be notified of
    progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc CatalystX::InjectComponent

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=CatalystX-InjectComponent>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/CatalystX-InjectComponent>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/CatalystX-InjectComponent>

    *   Search CPAN

        <http://search.cpan.org/dist/CatalystX-InjectComponent/>

ACKNOWLEDGEMENTS
COPYRIGHT & LICENSE
    Copyright 2009 Robert Krimen, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
0.024 Wednesday October 14 18:30:13 PDT 2009:
    - Make sure we're being used with Moose-based Catalyst

0.023 Monday October 12 16:10:13 PDT 2009
    - Added Catalyst::Runtime dependency (thanks Andreas Koenig)

0.022 Wednesday October 07 22:22:18 PDT 2009:
    - Remove plugins from tests to fix dependency failure (thanks bobtfish)

0.021 Wednesday June 24 00:49:08 PDT 2009:
    - Added parent dependency

0.020 Wednesday June 10 11:24:59 PDT 2009:
    - Altered injection interface: into =>, component =>, as =>

0.010 Monday June 08 16:08:08 PDT 2009:
    - Initial release


t/Test/Catalyst.pm
package t::Test::Catalyst;

use strict;
use warnings;

use Catalyst::Runtime '5.70';
use CatalystX::InjectComponent;

use Moose;
BEGIN { extends qw/Catalyst/ }

use Catalyst qw/-Debug
                ConfigLoader
                Static::Simple/;
our $VERSION = '0.01';

__PACKAGE__->config( name => 't::Test::Catalyst' );

after 'setup_components' => sub {
    my $self = shift;
    CatalystX::InjectComponent->_inject_component(
        __PACKAGE__,
        't::Test::Component1',
        'Controller::Component1',
#        into => __PACKAGE__,
#        base => 'SuperFunComponent',
#        moniker => 'Controller::SuperFun',
    );
};

__PACKAGE__->setup();

1;


t/Test/Component1.pm
package t::Test::Component1;

use strict;
use warnings;

use parent qw/Catalyst::Controller/;

sub default :Path {
    my ( $self, $ctx ) = @_;
}

sub durp :Local {
}

1;


t/Test/Catalyst/Controller/Root.pm
package t::Test::Catalyst::Controller::Root;

use strict;
use warnings;
use parent 'Catalyst::Controller';

__PACKAGE__->config->{namespace} = '';

sub default :Path {
    my ( $self, $c ) = @_;
    $c->response->body( 'Page not found' );
    $c->response->status(404);
}

sub end : ActionClass('RenderView') {}

1;


Makefile.PL
use inc::Module::Install;

name     'CatalystX-InjectComponent';
all_from 'lib/CatalystX/InjectComponent.pm';
author   'Robert Krimen <rkrimen@cpan.org>';
license  'perl';

build_requires 'Test::More';

resources repository => 'http://github.com/robertkrimen/CatalystX-InjectComponent/tree/master';

{
    require ExtUtils::MakeMaker;
    use strict;
    no strict 'refs';

    my $libscan = \&{"ExtUtils::MM_Any::libscan"};
    *{"ExtUtils::MM_Any::libscan"} = sub {
        return '' unless $libscan->(@_);
        return '' if $_[1] =~ /\.sw[p-z]$/;
        return $_[1];
    };
}

{
    map { my ($pk, $vr) = split m/\s/; build_requires $pk => $vr || 0 } grep { ! /^\s*#/ } split m/\n/, <<_END_;
Test::Most
_END_

    map { my ($pk, $vr) = split m/\s/; requires $pk => $vr || 0 } grep { ! /^\s*#/ } split m/\n/, <<_END_;
parent
Carp
Class::Inspector
Devel::InnerPackage
Catalyst::Runtime 5.8
_END_
}

if (-e 'inc/.author') {
    my $all_from = join '/', 'lib', split m/-/, name . '.pm';
    `perldoc -tF $all_from > README` if ! -e 'README' || (stat $all_from)[9] > (stat 'README')[9];
}


WriteAll;



lib/CatalystX/InjectComponent.pm
package CatalystX::InjectComponent;

use warnings;
use strict;

=head1 NAME

CatalystX::InjectComponent - Inject components into your Catalyst application

=head1 VERSION

Version 0.024

=cut

our $VERSION = '0.024';

=head1 SYNOPSIS

    package My::App;

    use Catalyst::Runtime '5.80';

    use Moose;
    BEGIN { extends qw/Catalyst/ }

    ...

    after 'setup_components' => sub {
        my $class = shift;
        CatalystX::InjectComponent->inject( into => $class, component => 'MyModel' );
        if ( $class->config->{ ... ) {
            CatalystX::InjectComponent->inject( into => $class, component => 'MyRootV2', as => 'Controller::Root' );
        }
        else {
            CatalystX::InjectComponent->inject( into => $class, component => 'MyRootV1', as => 'Root' ); # Controller:: will be automatically prefixed
        }
    };

=head1 DESCRIPTION

CatalystX::InjectComponent will inject Controller, Model, and View components into your Catalyst application at setup (run)time. It does this by creating
a new package on-the-fly, having that package extend the given component, and then having Catalyst setup the new component (via C<< ->setup_component >>)

=head1 So, how do I use this thing?

You should inject your components when appropiate, typically after C<setup_compenents> runs

If you're using the Moose version of Catalyst, then you can use the following technique:

    use Moose;
    BEGIN { extends qw/Catalyst/ }

    after 'setup_components' => sub {
        my $class = shift;

        CatalystX::InjectComponent->inject( into => $class, ... )
    };

=head1 METHODS

=head2 CatalystX::InjectComponent->inject( ... )

    into        The Catalyst package to inject into (e.g. My::App)
    component   The component package to inject
    as          An optional moniker to use as the package name for the derived component 

For example:

    ->inject( into => My::App, component => Other::App::Controller::Apple )
        
        The above will create 'My::App::Controller::Other::App::Controller::Apple'

    ->inject( into => My::App, component => Other::App::Controller::Apple, as => Apple )

        The above will create 'My::App::Controller::Apple'

=head1 ACKNOWLEDGEMENTS

Inspired by L<Catalyst::Plugin::AutoCRUD>

=cut

use Devel::InnerPackage;
use Class::Inspector;
use Carp;

sub put_package_into_INC ($) {
    my $package = shift;
    (my $file = "$package.pm") =~ s{::}{/}g;
    $INC{$file} ||= 1;
}

sub loaded ($) {
    my $package = shift;
    if ( Class::Inspector->loaded( $package ) ) {
        put_package_into_INC $package; # As a courtesy
        return 1;
    }
    return 0;
}

sub inject {
    my $self = shift;
    my %given = @_;

    my ($into, $component, $as);
    if ( $given{catalyst} ) { # Legacy argument parsing
        ($into, $component, $as) = @given{ qw/catalyst component into/ };
    }
    else {
        ($into, $component, $as) = @given{ qw/into component as/ };
    }
    
    croak "No Catalyst (package) given" unless $into;
    croak "No component (package) given" unless $component;

    unless ( loaded $component ) {
        eval "require $component;" or croak "Couldn't require (component base) $component: $@";
    }

    $as ||= $component;
    unless ( $as =~ m/^(?:Controller|Model|View)::/ || $given{skip_mvc_renaming} ) {
        my $category;
        for (qw/ Controller Model View /) {
            if ( $component->isa( "Catalyst::$_" ) ) {
                $category = $_;
                last;
            }
        }
        croak "Don't know what kind of component \"$component\" is" unless $category;
        $as = "${category}::$as";
    }
    my $component_package = join '::', $into, $as;

    unless ( loaded $component_package ) {
        eval "package $component_package; use parent qw/$component/; 1;" or
            croak "Unable to build component package for \"$component_package\": $@";
        put_package_into_INC $component_package; # As a courtesy
    }

    $self->_setup_component( $into => $component_package );
    for my $inner_component_package ( Devel::InnerPackage::list_packages( $component_package ) ) {
        $self->_setup_component( $into => $inner_component_package );
    }
}

sub _setup_component {
    my $self = shift;
    my $into = shift;
    my $component_package = shift;
    $into->components->{$component_package} = $into->setup_component( $component_package );
}

=head1 AUTHOR

Robert Krimen, C<< <rkrimen at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-catalystx-injectcomponent at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CatalystX-InjectComponent>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc CatalystX::InjectComponent


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=CatalystX-InjectComponent>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/CatalystX-InjectComponent>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/CatalystX-InjectComponent>

=item * Search CPAN

L<http://search.cpan.org/dist/CatalystX-InjectComponent/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 Robert Krimen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of CatalystX::InjectComponent
